### PR TITLE
Accept Trailing Slashes and base paths in nested configurations

### LINKED ISSUE #1: Wrong optimistic matching of route with undesired pattern
``` js
const routes = {
  '/': h('h1', 'Hello, world'),
  '/authors': h('h1', 'Authors'),
  '/books': h('h1', 'Books'),
  '/books/:id': id => h('h1', `BookId: ${id}`)
}
```

When I navigated to `/home/33/books/10` I was given `/books/33`

https://github.com/cyclejs/cycle-core/issues/159#issuecomment-143846923

### PR DESCRIPTION
**Pull request by TylorS**
I'm not sure if this is how you would like to handle this, but it's a simple solution but it also removes 'optimistic matching'. Which I have preference for, but you may not see desire for that.

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -51,4 +75,16 @@ describe('switchPath', () => {
...
+    const {path, value} = switchPath('/1736', {
+      '/bar': 123,
+      '/:id': {
+        '/': id => `id is ${id}`,
+        '/home': 789
+      }
+    });
+    expect(path).to.be.equal('/1736');
+    expect(value).to.be.equal('id is 1736');
+  })
```
Comment (staltz) [inline test/index.js]: Missing semicolon here.

File: test/index.js [hunk 2]
```diff
@@ -51,4 +75,16 @@ describe('switchPath', () => {
     expect(path).to.be.equal('/home/1736');
     expect(value).to.be.equal('id is 1736');
   });
+
```
Comment (staltz) [inline test/index.js:78]: (Below) Good test. I agree switchPath should handle that.

File: test/index.js [hunk 3]
```diff
@@ -19,16 +31,28 @@ describe('switchPath', () => {
...
+  });
+
+  it('should match a path with an extra trailing slash', () => {
+    const {path, value} = switchPath('/home/foo/', {
+      '/bar': 123,
+      '/home': {
+        '/foo': 456
+      }
+    });
+    expect(path).to.be.equal('/home/foo');
```
Comment (staltz) [inline test/index.js:55]: :+1:

File: test/index.js [hunk 4]
```diff
@@ -19,16 +31,28 @@ describe('switchPath', () => {
...
-  it('should match a path on an incomplete pattern', () => {
+  it('should NOT match a path on an incomplete pattern', () => {
     const {path, value} = switchPath('/home/foo', {
       '/bar': 123,
       '/home': 456,
     });
-    expect(path).to.be.equal('/home');
+    expect(path).to.be.equal(null);
+    expect(value).to.be.equal(null);
+  });
```
Comment (staltz) [inline test/index.js:46]: I really can't accept this test, this is the opposite direction of what I want switchPath to accomplish. Optimistically matching a path based on a prefix is super important for how I want to use it for nested dialogues in Cycle. The case being: suppose `fullPath = '/movies/1239'` and imagine the parent using switchPath on

``` js
const {path, Component} = switchPath(fullPath, {
  '/about': AboutComponent,
  '/profile': ProfileComponent,
  '/movies': MoviesComponent,
}
```

SwitchPath returns the right hand side as `Component`, which is a dialogue function.
Then the parent calls `Component` and passes to it both the `fullPath` and `path`. Then the Component (for instance `MoviesComponent` removes `path` as a prefix of `fullPath` to get `subPath`, and internally the child Component will do another switchPath on that subPath:

``` js
const {path, vtree} = switchPath(subPath, {
  '/': moviesHomeVTree,
  '/:id': movieDetailVTree,
})
```

See the importance of that feature now?

File: test/index.js [hunk 5]
```diff
@@ -12,6 +12,18 @@ describe('switchPath', () => {
...
+      '/bar': 123,
+      '/home': {
+        '/': 456,
+        '/foo': 789
+      }
+    });
+    expect(path).to.be.equal('/home');
+    expect(value).to.be.equal(456);
+  });
+
```
Comment (staltz) [inline test/index.js:26]: :+1:

File: test/index.js (hunk 4 above)
Comment (TylorS) [inline test/index.js:46]: Yes I do :+1: 

Reverted

Comment (staltz) [inline test/index.js:46]: But I'd like to merge the other features. Can you somehow revert changes just for this test to behave like previously?

Comment (TylorS) [inline test/index.js:46]: Already done.

**Comment by staltz** [comment]
Perfect PR. Big thanks @TylorS

**Pull request merged**